Add unit tests for ProductRepository

nima-karimi opened the pull request:
Changes proposed in this Pull Request:

This PR adds unit tests for the `ProductRepository` class. It also removes unused functions from the ProductRepository class. These methods were previously added here but I noticed that they are no longer used anywhere.

I was thinking of deprecating them first but not sure if it's worth the effort since they are very specific methods and unlikely used by any third parties.

Continuing from #846

### Detailed test instructions:

- Run `phpunit` and make sure all the tests pass

nima-karimi committed:
Use array diff to remove `variation` type

nima-karimi committed:
Remove ProductRepository's unused methods

nima-karimi committed:
Add unit tests for `ProductRepository`

mikkamp commented on tests/Unit/Product/ProductRepositoryTest.php:
```diff
@@ -0,0 +1,272 @@
+
+		$results = $this->product_repository->find();
+
+		$variations = array_map( 'wc_get_product', $variable_product->get_children() );
+		$this->assertCount( count( $variations ) + 2, $results );
+		$this->assertEqualSets(
+			array_merge(
+				[
+					wc_get_product( $simple_product->get_id() ),
+					wc_get_product( $variable_product->get_id() ),
+				],
+				$variations
+			),
+			$results
+		);
```
Works fine although I would have refactored it slightly to create something like:
```php
$expected = array_merge(
	[
		wc_get_product( $simple_product->get_id() ),
		wc_get_product( $variable_product->get_id() ),
	],
	array_map( 'wc_get_product', $variable_product->get_children() )
);
```

Then make the assertCount just be a `count( $expected )`. At first glance the variations + 2 isn't clear what you are counting.

nima-karimi replied:
Changed in https://github.com/woocommerce/google-listings-and-ads/pull/855/commits/cde5c82ce5d8b28d867b19646a72d2724df21649

mikkamp commented on tests/Unit/Product/ProductRepositoryTest.php:
```diff
@@ -0,0 +1,272 @@
+		$this->assertEquals( [ wc_get_product( $product_1 ) ], $this->product_repository->find( $query_args ) );
+	}
+
+	public function test_find_by_ids() {
```
In this test we only create two products, so `find` would be equal to `find_by_ids`. Do we want to add a few more products to make sure we only get back the ones we requested?

nima-karimi replied:
Good call. Added in https://github.com/woocommerce/google-listings-and-ads/pull/855/commits/1dc146f57235c45fafe6dc1af524f42f292f0c09

mikkamp commented on tests/Unit/Product/ProductRepositoryTest.php:
```diff
@@ -0,0 +1,272 @@
+		// a variable product that is sync ready along with all its variations
+		$variable_product = WC_Helper_Product::create_variation_product();
+		$this->product_meta->update_visibility( wc_get_product( $variable_product ), ChannelVisibility::SYNC_AND_SHOW );
+		foreach ( $variable_product->get_children() as $variation_id ) {
+			$this->product_meta->update_visibility( wc_get_product( $variation_id ), ChannelVisibility::SYNC_AND_SHOW );
+		}
```
As far as I know we don't set this for each variation when updating a product through the ChannelVisibilityMetaBox. Although we might want to consider doing so, as it can lead to creating incomplete batches when syncing products. Was still planning to log an issue for this.

mikkamp replied:
Here is the issue: https://github.com/woocommerce/google-listings-and-ads/issues/861

mikkamp commented on tests/Unit/Product/ProductRepositoryTest.php:
```diff
@@ -0,0 +1,272 @@
+
+		$results = $this->product_repository->find_sync_ready_products();
+
+		// compare the IDs because the objects might not be identical
```
Why are we able to compare the objects in all the other tests? Is there something specific that's making them different for this test?

nima-karimi replied:
For some reason, the `internal_meta_keys` property of the `data_store` instance embedded within the `WC_Product` object gets modified here in this particular function. 

It might have something to do with the extra call to `filter_sync_ready_products`, which calls several methods of the `ProductMetaHandler` class, which in turn leads to a call to the WC internal function `maybe_read_meta_data` on the product object. This method is probably adding all those extra data into the product object, making it different than what we expect.

I went with comparing the objects in other tests but generally, I think it's just safer to compare the IDs. Since we don't care about the actual ORM object here.

![image](https://user-images.githubusercontent.com/73110514/124655189-94fab200-deb0-11eb-86b4-1bd8e37b7a13.png)

![image](https://user-images.githubusercontent.com/73110514/124655215-9d52ed00-deb0-11eb-9ab7-60d621ae121b.png)

nima-karimi replied:
I modified all tests in this test case to only compare product IDs: https://github.com/woocommerce/google-listings-and-ads/pull/855/commits/73a3d5768139921218d4c1b2f964d677887197de

mikkamp replied:
Thanks, that's probably more foolproof further down the line anyways.

mikkamp reviewed (commented):
I think it's fine to remove those functions without deprecating them as they were internally used and only got unused because of all the refactoring we did.

I left a few small comments on some of the tests.

nima-karimi committed:
Use a variable to store expected value of the test

nima-karimi committed:
Create a few more products to test `find_by_ids`

nima-karimi commented:
Thanks for the review, @mikkamp . I made some changes and left a comment regarding the ID comparison. Ready for another review.

nima-karimi committed:
Add helper method to get product's ID

nima-karimi committed:
Compare the product's ID instead of object for more accurate results

The product objects might be modified by WooCommerce halfway through execution and in the ProductRepositoryTest it's safer to compare their IDs instead of the objects. Test results would be the same because we only expect the methods to return the product we requested (based on our query) and we don't care about the object that is returned or what it contains (that's up to WooCommerce internals).

mikkamp reviewed (approved):
Thanks for all the changes, those look good and all the tests are still passing.
